Create optimize api tutorial and revise admin api tutorial (#3708)
camunda/camunda-docs: optimize/apis-tools/optimize-api/tutorial.md

---
id: optimize-api-tutorial
title: Tutorial
description: "Step through an example to view your existing clients, create a client, view a particular client's details, and delete a client."
---

In this tutorial, we'll step through examples to highlight the capabilities of the Optimize API, such as listing your existing dashboard IDs, or deleting a dashboard.

## Getting started

- You need authentication to access the API endpoints. Find more information [here](./optimize-api-authentication.md).
- To properly execute the commands to list existing dashboard IDs, and delete a dashboard, ensure you have [created a collection](../../components/userguide/collections-dashboards-reports.md) containing a [dashboard](../../components/userguide/creating-dashboards.md).

## Prerequisites

> - Ensure you have [Node.js](https://nodejs.org/en/download) downloaded.

Author: Adjusted in latest commit 👍

- If you haven't done so already, [create a cluster]($docs$/guides/assets/react-components/create-cluster).

> - Upon cluster creation, [create your first client]($docs$/guides/setup-client-connection-credentials). Ensure you determine the scoped access for client credentials, and check all the boxes for Zeebe client scopes.

Author: Resolved in commit 👍

<!--- I can't recall, does it matter how this is set up, or solely the scope? --->

:::note
Make sure you keep the generated client credentials in a safe place. The **Client secret** will not be shown again. For your convenience, you can also download the client information to your computer.
:::

- In this tutorial, we utilize a JavaScript-written [GitHub repository](https://github.com/camunda/camunda-api-tutorials) to write and run requests. Clone this repo before getting started.
- Run `npm install` to ensure you have updated dependencies.

## Set up authentication

To get started, examine the `auth.js` file in the GitHub repository. This file contains a function named `getAccessToken` which executes an OAuth 2.0 protocol to retrieve authentication credentials based on your client id and client secret. We will call this function whenever we need an authentication token for an API request.

1. To set up your credentials, create an `.env` file which will be protected by the `.gitignore` file. These keys will be consumed by the `auth.js` file to execute the OAuth protocol, and should be saved when you generate your client credentials in [prerequisites](#prerequisites).

> 2. Examine the existing `.env.example` file for an example of how your `.env` file should look upon completion. You will need to add your `COMPONENTS_CLIENT_ID`, `COMPONENTS_CLIENT_SECRET`, `OPTIMIZE_BASE_URL`, and `OPTIMIZE_AUDIENCE`, which is `https://optimize.camunda.io` in a Camunda 8 SaaS environment.

Author: Resolved in commit 👍

:::note

In this tutorial, we will execute arguments to list existing dashboard IDs and delete a dashboard. You can examine the framework for processing these arguments in the `cli.js` file before getting started.

:::

## GET a list of existing dashboard IDs

First, let's script an API call to list our existing dashboard IDs.

To do this, take the following steps:

1. Examine the function `async function listDashboards([collectionId])` at the top of the `optimize.js` file. This is where you will script out your API call.
2. Within the function, you must first apply an access token for this request:

```
const accessToken = await getAccessToken("components", optimizeAudience);
```

> 3. Using your generated client credentials from [prerequisites](#prerequisites), call your Optimize audience and base URL. Ensure these credentials are added to your `.env` file, and script the references to the values. For example:

Reviewer: ```suggestion
3. Using your generated client credentials from [prerequisites](#prerequisites), capture your Optimize audience and base URL. Ensure these credentials are added to your `.env` file, and retrieve the values from the process environment:
```

Non-blocking suggestion, for precision.

Author: Resolved in commit 👍

```
const optimizeAudience = process.env.OPTIMIZE_AUDIENCE;
const optimizeApiUrl = process.env.OPTIMIZE_BASE_URL;
```

4. Script the API endpoint to list the dashboard IDs of a particular collection:

```
const url = `${optimizeApiUrl}/api/public/dashboard?collectionId=${collectionId}`;
```

5. Configure your GET request to the appropriate endpoint, including an authorization header based on the previously acquired `accessToken`:

```
  const options = {
    method: "GET",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

6. Call the collection's endpoint, process the results from the API call, emit the dashboard IDs to output, and emit an error message from the server if necessary:

```
  try {
    const response = await axios(options);
    const results = response.data;

    results.forEach(x => console.log(`ID: ${x.id}`));
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
```

> 7. In your terminal, run `node cli.js optimize list` for a list of your existing dashboard IDs.

Reviewer: non-blocking question: 

Should we show example output here?

Author: Added in latest commit 👍

:::note
This `list` command is connected to the `listDashboards` function at the bottom of the `optimize.js` file, and executed by the `cli.js` file. While we will view dashboard IDs and delete a dashboard in this tutorial, you may add additional arguments depending on the API calls you would like to make.
:::

If you have any existing dashboards, the `ID: ${x.id}` will now output. If you have an invalid API name or action name, or no arguments provided, or improper/insufficient credentials configured, an error message will output as outlined in the `cli.js` file.

## DELETE a dashboard

To delete a dashboard, take the following steps:

> 1. Outline your function, similar to the steps above:

Reviewer: non-blocking question:

is it worth noting that the last line of this block is different, intentionally, because we'll be hitting a different endpoint this time?

Author: Added in latest commit 👍

```
async function deleteDashboard([dashboardId]) {
console.log(`deleting dashboard ${dashboardId}`);

const optimizeAudience = process.env.OPTIMIZE_AUDIENCE;
const accessToken = await getAccessToken("components", optimizeAudience);
const optimizeApiUrl = process.env.OPTIMIZE_API_URL;

const url = `${optimizeApiUrl}/public/dashboard/${dashboardId}`;
```

2. Configure the API call using the DELETE method:

```
  const options = {
    method: "DELETE",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

3. Process the results from the API call. For example:

```
  try {
    // Call the delete endpoint.
    const response = await axios(options);

    // Process the results from the API call.
    if (response.status === 204) {
      console.log(`Dashboard ${clientId} was deleted!`);
    } else {
      // Emit an unexpected error message.
      console.error("Unable to delete dashboard!");
    }
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
}
```

> 4. In your terminal, run `node cli.js optimize delete` to delete your dashboard.

Reviewer: non-blocking question:

again, should we show example output? (In this case it's just the text "dashboard was deleted!" so I think it's less valuable than showing example output on the first method. But I'm still asking the question.)

Author: Added in latest commit 👍

- Ensure you have [Node.js](https://nodejs.org/en/download) installed as this will be used for methods that can be called by the CLI (outlined later in this guide). Run `npm install` to ensure you have updated dependencies. Run `npm install` to ensure you have updated dependencies.

## Set up authentication

To get started, examine the `auth.js` file in the GitHub repository. This file contains a function named `getAccessToken` which executes an OAuth 2.0 protocol to retrieve authentication credentials based on your client id and client secret. We will call this function whenever we need an authentication token for an API request.

:::note
You may notice in the `auth.js` file that `getAccessToken` takes two arguments. This separates the authentication between the Administration API and the other component APIs (in this case, the component being Optimize).
:::

1. To set up your credentials, create an `.env` file which will be protected by the `.gitignore` file. These keys will be consumed by the `auth.js` file to execute the OAuth protocol, and should be saved when you generate your client credentials in [prerequisites](#prerequisites).

> 2. Examine the existing `.env.example` file for an example of how your `.env` file should look upon completion. You will need to add your `COMPONENTS_CLIENT_ID`, `COMPONENTS_CLIENT_SECRET`, `OPTIMIZE_BASE_URL`, and `OPTIMIZE_AUDIENCE`, which is `optimize.camunda.io` in a Camunda 8 SaaS environment. Do not place your credentials in the `.env.example` file, as this example file is not protected by the `.gitignore`.

Author: > and `OPTIMIZE_AUDIENCE`, which is `optimize.camunda.io`

@toco-cam I had an `https://` before this in the example env file in the tutorial, which was causing this to fail. I fixed this in the repo, and it works for me now. Let me know if you're still having issues, and thanks for catching!

Author: Oops. I published this before I saw the message from @pepopowitz above 😅 

Regardless, let us know if you're still having trouble!

Author: We won't have any users working in an ultrawombat environment, so I do wonder if your issues are coming from your .env file -- specifically, Steve mentioned your client ID/secrets may no longer be active.

If you try our tutorial from a personal setup at camunda.io with updated cluster/client details, do you run into the same issues?

Reviewer 2: @christinaausley: Ok, it was most likely the ultrawombat environment. I got it working on https://bru-2.optimize.camunda.io.

But the error message was strange - it seemed as Ultrawombat is using different text encoding.

Reviewer: I suspect the text encoding message is more a symptom of the issue, than the issue itself. Those types of messages often occur when an HTML error page is being served and the caller tries to parse it as JSON.

Here's what I think was happening:

* You created the API client in the ultrawombat environment
* The "authentication" endpoint [is hardcoded in the tutorial](https://github.com/camunda/camunda-api-tutorials/blob/ec7e97a91d911344c9898e275c07917a71a68e89/auth.js#L41) to the production SaaS auth URL
* You were getting an error page when attempting to authenticate, because the API keys weren't known to the production environment

This does raise the question -- should the auth URL be hardcoded in the sample code, or should it be configurable by an environment variable?

```
const optimizeAudience = process.env.OPTIMIZE_AUDIENCE;
const optimizeApiUrl = process.env.OPTIMIZE_BASE_URL;
```

4. Script the API endpoint to list the dashboard IDs of a particular collection:

```
const url = `${optimizeApiUrl}/api/public/dashboard?collectionId=${collectionId}`;
```

5. Configure your GET request to the appropriate endpoint, including an authorization header based on the previously acquired `accessToken`:

```
  const options = {
    method: "GET",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

6. Call the collection's endpoint, process the results from the API call, emit the dashboard IDs to output, and emit an error message from the server if necessary:

```
  try {
    const response = await axios(options);
    const results = response.data;

    results.forEach(x => console.log(`ID: ${x.id}`));
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
```

> 7. In your terminal, run `node cli.js optimize list` for a list of your existing dashboard IDs. If you have any existing dashboards within a collection, you will see an output similar to the following:

Reviewer 3: I needed to provide a collectionId here. Is that intentional? 

If I don't interpret step 4 correctly, I miss this and don't get an error or anything to guide me forward.

Also, I didn't see anything about how to find the collectionId. That should be added.

Author: Yes! This would be a good place to note that, because I don't think @toco-cam anticipated that either. Will add into a commit tomorrow 👍

```
async function deleteDashboard([dashboardId]) {
console.log(`deleting dashboard ${dashboardId}`);

const optimizeAudience = process.env.OPTIMIZE_AUDIENCE;
const accessToken = await getAccessToken("components", optimizeAudience);
const optimizeApiUrl = process.env.OPTIMIZE_API_URL;

const url = `${optimizeApiUrl}/public/dashboard/${dashboardId}`;
```

2. Configure the API call using the DELETE method:

```
  const options = {
    method: "DELETE",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

3. Process the results from the API call. For example:

```
  try {
    // Call the delete endpoint.
    const response = await axios(options);

    // Process the results from the API call.
    if (response.status === 204) {
      console.log(`Dashboard ${clientId} was deleted!`);
    } else {
      // Emit an unexpected error message.
      console.error("Unable to delete dashboard!");
    }
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
}
```

> 4. In your terminal, run `node cli.js optimize delete` to delete your dashboard. You will see a response similar to the following:

Reviewer 3: Was happy to see an error here if I don't provide a `dashboardId`. But after I did, I'm still getting an ECONNREFUSED when I run the completed file. 

If I step through what's written here, it doesn't work, but I also don't get an error.

I incorporated the changes @pepopowitz made in https://github.com/camunda/camunda-api-tutorials/pull/7/files (adding `/api` to the URL changing the variable in the output to dashboardId). I don't get an error.

Author: Okay -- I haven't analyzed the changes from @pepopowitz with the new auth layout into this PR. I was anticipating launching a new PR with the updated auth changes across administration, optimize, and modeler.

I'm thinking it might be best to move this back into `draft`, and I can adjust the auth references across Administration and Optimize in this PR, with a fresh PR for Modeler.

Reviewer: There are several bug fixes that I snuck into https://github.com/camunda/camunda-api-tutorials/pull/7, which I should have opened a separate PR for. They were all related to the `delete` functionality.

To do this, take the following steps:

1. In the file named `optimize.js`, outline the authentication and authorization configuration in the first few lines. This will pull in your `.env` variables to obtain an access token before making any API calls:

```
const authorizationConfiguration = {
  clientId: process.env.OPTIMIZE_CLIENT_ID,
  clientSecret: process.env.OPTIMIZE_CLIENT_SECRET,
  audience: process.env.OPTIMIZE_AUDIENCE
};
```

2. Examine the function `async function listDashboards([collectionId])` below this configuration. This is where you will script out your API call.
3. Within the function, you must first apply an access token for this request, so your function should now look like the following:

```
async function listDashboards([collectionId]) {
  const accessToken = await getAccessToken(authorizationConfiguration);
}
```

4. Using your generated client credentials from [prerequisites](#prerequisites), capture your Optimize API URL beneath your call for an access token by defining `optimizeApiUrl`:

`const optimizeApiUrl = process.env.OPTIMIZE_BASE_URL;`

5. On the next line, script the API endpoint to list your existing dashboard IDs for a particular collection:

> ```

Reviewer: ```suggestion
```javascript
```

const url = `${optimizeApiUrl}/api/public/dashboard?collectionId=${collectionId}`;
```

6. Configure your GET request to the appropriate endpoint, including an authorization header based on the previously acquired `accessToken`:

> ```

Reviewer: ```suggestion
```javascript
```

  const options = {
    method: "GET",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

7. Call the collection's endpoint, process the results from the API call, emit the dashboard IDs to output, and emit an error message from the server if necessary:

> ```

Reviewer: ```suggestion
```javascript
```

try {
  const response = await axios(options);
  const results = response.data;

  results.forEach((x) => console.log(`ID: ${x.id}`));
} catch (error) {
  // Emit an error from the server.
  console.error(error.message);
}
```

8. In your terminal, run `node cli.js optimize list <collection ID>`, where `<collection ID>` is where you can paste the ID of your collection for a list of your existing dashboard IDs within this particular collection. If you have any existing dashboards within a collection, you will see an output similar to the following:

`ID: 12345`

:::note
This `list` command is connected to the `listDashboards` function at the bottom of the `optimize.js` file, and executed by the `cli.js` file. While we will view dashboard IDs and delete a dashboard in this tutorial, you may add additional arguments depending on the API calls you would like to make.
:::

If you have any existing dashboards, the `ID: ${x.id}` will now output. If you have an invalid API name or action name, or no arguments provided, or improper/insufficient credentials configured, an error message will output as outlined in the `cli.js` file.

## DELETE a dashboard

> To delete a dashboard, take the following steps:

Reviewer: ```suggestion
To delete a dashboard, capture its ID from the previous exercise and take the following steps:
```

@christinaausley what do you think about this, re: https://github.com/camunda/camunda-docs/pull/3708/files#r1607280045 ???

Author: AH now I see what you're saying! I'll add and backport, but will request one more green check 👍

1. Outline your function, similar to the steps above. Note that the URL endpoint will look different, as you are accessing a different endpoint in this request (using a dashboard ID) than in the prior request (using a collection ID):

> ```

Reviewer: ```suggestion
```javascript
```

> 1. Outline your function, similar to the steps above. Note that the URL endpoint will look different, as you are accessing a different endpoint in this request (using a dashboard ID) than in the prior request (using a collection ID):

Author: We have:

> In your terminal, run `node cli.js optimize list <collection ID>`, where `<collection ID>` is where you can paste the ID of your collection for a list of your existing dashboard IDs within this particular collection. If you have any existing dashboards within a collection, you will see an output similar to the following:

But it sounds like this is not enough. Can you propose the change you're looking for, and I can review/backport?

async function deleteDashboard([dashboardId]) {
  console.log(`deleting dashboard ${dashboardId}`);

  const accessToken = await getAccessToken(authorizationConfiguration);

  const optimizeApiUrl = process.env.OPTIMIZE_BASE_URL;
  const url = `${optimizeApiUrl}/api/public/dashboard/${dashboardId}`;
```

2. Configure the API call using the DELETE method:

> ```

Reviewer: ```suggestion
```javascript
```

  const options = {
    method: "DELETE",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

3. Process the results from the API call. For example:

> ```

Reviewer: ```suggestion
```javascript
```
